resolver/resolver.cc

```diff
@@ -1576,6 +1624,172 @@ class ResolveTypeMembersAndFieldsWalk {
         }
     }
 
+    void validateNonForcingIsA(core::Context ctx, const ast::Send &send) {
+        constexpr string_view method = "T::NonForcingConstants.non_forcing_is_a?";
+
+        if (send.numPosArgs != 2) {
+            return;
+        }
+
+        auto [posEnd, kwEnd] = send.kwArgsRange();
+        auto numKwArgs = (kwEnd - posEnd) >> 1;
+        if (numKwArgs > 1) {
+            return;
+        }
+
+        auto stringLoc = send.args[1].loc();
+
+        auto *literalNode = ast::cast_tree<ast::Literal>(send.args[1]);
+        if (literalNode == nullptr) {
+            if (auto e = ctx.beginError(stringLoc, core::errors::Resolver::LazyResolve)) {
+                e.setHeader("`{}` only accepts string literals", method);
+            }
+            return;
+        }
+
+        if (!core::isa_type<core::LiteralType>(literalNode->value)) {
+            if (auto e = ctx.beginError(stringLoc, core::errors::Resolver::LazyResolve)) {
+                e.setHeader("`{}` only accepts string literals", method);
+            }
+            return;
+        }
+
+        auto literal = core::cast_type_nonnull<core::LiteralType>(literalNode->value);
+        if (literal.literalKind != core::LiteralType::LiteralTypeKind::String) {
+            // Infer will report a type error
+            return;
+        }
+
+        core::TypePtr packageType = nullptr;
+        optional<core::LocOffsets> packageLoc;
+        if (send.hasKwArgs()) {
+            // this means we got the third package arg
+            auto *key = ast::cast_tree<ast::Literal>(send.args[posEnd]);
+            if (!key || !key->isSymbol(ctx) || key->asSymbol(ctx) != ctx.state.lookupNameUTF8("package")) {
```

Comment:
Dang this should never have been written this way—we should have `"package"` in `generate_names.cc`.

Reply:
(Note: I didn't write this; I'm just moving it.)

Reply:
I can spruce this up in a subsequent PR.

Reply:
yep, i should have commented on the red part of the diff to make that clear